blockchain/blockchain.go

```diff
@@ -0,0 +1,316 @@
+		},
+		reqPool: &sync.Pool{
+			New: func() interface{} {
+				return new(http.Request)
+			},
+		},
+	}
+	if bl.authorizer != "" {
+		if err := bl.SetAuth(context.Background(), h.ID(), bl.authorizer, true); err != nil {
+			return nil, err
+		}
+	}
+	return bl, nil
+}
+
+func (bl *FxBlockchain) Start(ctx context.Context) error {
+	listen, err := gostream.Listen(bl.h, FxBlockchainProtocolID)
+	if err != nil {
+		return err
+	}
+	bl.s.Handler = http.HandlerFunc(bl.serve)
+	go func() { bl.s.Serve(listen) }()
+	return nil
+}
+
+func (bl *FxBlockchain) putBuf(buf *bytes.Buffer) {
+	buf.Reset()
+	bl.bufPool.Put(buf)
+}
+func (bl *FxBlockchain) putReq(req *http.Request) {
+	*req = http.Request{}
+	bl.reqPool.Put(req)
+}
+
+func (bl *FxBlockchain) callBlockchain(ctx context.Context, action string, p interface{}) ([]byte, error) {
+	method := http.MethodPost
+	addr := "http://" + bl.blockchainEndPoint + "/" + strings.Replace(action, "-", "/", -1)
+
+	// Use the bufPool and reqPool to reuse bytes.Buffer and http.Request objects
+	buf := bl.bufPool.Get().(*bytes.Buffer)
+	req := bl.reqPool.Get().(*http.Request)
+	defer func() {
+		bl.putBuf(buf)
+		bl.putReq(req)
+	}()
+
+	if err := json.NewEncoder(buf).Encode(p); err != nil {
+		return nil, err
+	}
+	req, err := http.NewRequestWithContext(ctx, method, addr, buf)
+	if err != nil {
+		return nil, err
+	}
+
+	req.Header.Set("Content-Type", "application/json")
+
+	resp, err := bl.ch.Do(req)
+
+	if err != nil {
+		return nil, err
+	}
+
+	defer resp.Body.Close()
+
+	var bufRes bytes.Buffer
+	_, err = io.Copy(&bufRes, resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	b := bufRes.Bytes()
+	switch {
+	case resp.StatusCode != http.StatusOK:
+		return nil, fmt.Errorf("unexpected response: %d %s", resp.StatusCode, string(b))
+	default:
+		return b, nil
+	}
+}
+
+func (bl *FxBlockchain) serve(w http.ResponseWriter, r *http.Request) {
+	from, err := peer.Decode(r.RemoteAddr)
+	if err != nil {
+		log.Debug("cannot parse remote addr as peer ID: %v", err)
+		http.Error(w, "", http.StatusBadRequest)
+		return
+	}
+	action := path.Base(r.URL.Path)
+	if !bl.authorized(from, action) {
+		log.Debug("rejected unauthorized request from %s for action %s", from, action)
+		http.Error(w, "", http.StatusUnauthorized)
+		return
+	}
+	// Define a map of functions with the same signature as handleAction
+	actionMap := map[string]func(peer.ID, http.ResponseWriter, *http.Request){
+		actionSeeded: func(from peer.ID, w http.ResponseWriter, r *http.Request) {
+			bl.handleAction(actionSeeded, from, w, r)
+		},
+		actionAuth: func(from peer.ID, w http.ResponseWriter, r *http.Request) {
+			bl.handleAuthorization(from, w, r)
+		},
+	}
+
+	// Look up the function in the map and call it
+	handleActionFunc, ok := actionMap[action]
+	if !ok {
+		http.Error(w, "", http.StatusNotFound)
+		return
+	}
+	handleActionFunc(from, w, r)
+}
+
+func (bl *FxBlockchain) handleAction(action string, from peer.ID, w http.ResponseWriter, r *http.Request) {
```

Comment:
We should have explicit handlers for each of the blockchain actions. Having a function that takes arbitrary string is not a good idea i don't think. I recommend changing this to handleSeeded and remove the parameter action.